lib/resourcebuilder/batch.go

```diff
@@ -45,40 +44,51 @@ func (b *jobBuilder) Do(ctx context.Context) error {
 	if b.modifier != nil {
 		b.modifier(job)
 	}
-	_, updated, err := resourceapply.ApplyJob(b.client, job)
-	if err != nil {
+	if _, _, err := resourceapply.ApplyJob(b.client, job); err != nil {
 		return err
 	}
-	if updated && b.mode != InitializingMode {
-		return WaitForJobCompletion(ctx, b.client, job)
+	if b.mode != InitializingMode {
+		_, err := checkJobHealth(ctx, b.client, job)
+		return err
 	}
 	return nil
 }
 
 // WaitForJobCompletion waits for job to complete.
 func WaitForJobCompletion(ctx context.Context, client batchclientv1.JobsGetter, job *batchv1.Job) error {
 	return wait.PollImmediateUntil(defaultObjectPollInterval, func() (bool, error) {
-		j, err := client.Jobs(job.Namespace).Get(job.Name, metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("error getting Job %s: %v", job.Name, err)
+		if done, err := checkJobHealth(ctx, client, job); err != nil {
+			klog.Error(err)
+			return false, nil
+		} else if !done {
+			klog.V(4).Infof("Job %s in namespace %s is not ready, continuing to wait.", job.ObjectMeta.Namespace, job.ObjectMeta.Name)
 			return false, nil
 		}
+		return true, nil
+	}, ctx.Done())
+}
 
-		if j.Status.Succeeded > 0 {
-			return true, nil
-		}
+// checkJobHealth returns an error if the job status is bad enough to block further manifest application.
+func checkJobHealth(ctx context.Context, client batchclientv1.JobsGetter, job *batchv1.Job) (bool, error) {
+	j, err := client.Jobs(job.Namespace).Get(job.Name, metav1.GetOptions{})
+	if err != nil {
+		return false, fmt.Errorf("error getting Job %s: %v", job.Name, err)
+	}
 
-		// Since we have filled in "activeDeadlineSeconds",
-		// the Job will 'Active == 0' iff it exceeds the deadline.
```

Comment:
might as well fix `iff`

Reply:
Expanded to "if and only if" with 100231e -> c2af13b, which also rebases onto master to resolve conflicts with #403.